pkg/utils/envconfig.go

```diff
@@ -67,33 +59,71 @@ func (e *EnvConfig) GetLogger() *zap.SugaredLogger {
 	return e.logger
 }
 
-func (e *EnvConfig) SetupMetrics(ctx context.Context) error {
+func (e *EnvConfig) SetupObservability(ctx context.Context) error {
+	cfg := &observability.Config{}
+	err := json.Unmarshal([]byte(e.ObservabilityConfigJson), cfg)
+	if err != nil {
+		return fmt.Errorf("failed to parse observability config from env: %w", err)
+	}
+
+	cfg = observability.MergeWithDefaults(cfg)
+
 	logger := e.GetLogger()
-	// Convert json metrics.ExporterOptions to metrics.ExporterOptions.
-	metricsConfig, err := metrics.JSONToOptions(e.MetricsConfigJson)
+
+	pprof := k8sruntime.NewProfilingServer(logger.Named("pprof"))
+
+	pprof.UpdateFromConfig(cfg.Runtime)
+
+	otelResource, err := resource.Default(e.component)
 	if err != nil {
-		return err
+		logger.Warnw("encountered an error creating otel resource, some attributes may not be set", zap.Error(err))
 	}
 
-	if metricsConfig != nil {
-		metricsConfig.Component = strings.ReplaceAll(e.component, "-", "_")
-		if err := metrics.UpdateExporter(ctx, *metricsConfig, logger); err != nil {
-			return err
-		}
-		// Check if metrics config contains profiling flag
-		if metricsConfig.ConfigMap != nil {
-			if enabled, err := profiling.ReadProfilingFlag(metricsConfig.ConfigMap); err == nil && enabled {
-				// Start a goroutine to server profiling metrics
-				logger.Info("Profiling enabled")
-				go func() {
-					server := profiling.NewServer(profiling.NewHandler(logger, true))
-					// Don't forward ErrServerClosed as that indicates we're already shutting down.
-					if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
-						logger.Errorw("Profiling server failed", zap.Error(err))
-					}
-				}()
-			}
-		}
+	meterProvider, err := metrics.NewMeterProvider(
+		ctx,
+		cfg.Metrics,
+		metric.WithResource(otelResource),
+	)
+	if err != nil {
+		logger.Warnw("encountered error setting up otel meter provider, falling back to noop default", zap.Error(err))
+		meterProvider = eventingotel.DefaultMeterProvider(ctx, otelResource)
+	}
+
+	otel.SetMeterProvider(meterProvider)
+
+	err = runtime.Start(
+		runtime.WithMinimumReadMemStatsInterval(cfg.Runtime.ExportInterval),
+	)
+	if err != nil {
+		logger.Warnw("failed to start runtime metrics, will not export them", zap.Error(err))
 	}
+
+	tracerProvider, err := tracing.NewTracerProvider(
+		ctx,
+		cfg.Tracing,
+		trace.WithResource(otelResource),
+	)
+	if err != nil {
+		logger.Warnw("encountered error setting up otel tracer provider, falling back to noop default", zap.Error(err))
+		tracerProvider = eventingotel.DefaultTraceProvider(ctx, otelResource)
+	}
+
+	otel.SetTextMapPropagator(tracing.DefaultTextMapPropagator())
+	otel.SetTracerProvider(tracerProvider)
+
+	e.meterProvider = meterProvider
+	e.traceProvider = tracerProvider
+
 	return nil
 }
+
+func (e *EnvConfig) ShutdownObservability(ctx context.Context) {
```

Comment:
I like having this encapsulated in a function, rather than raw in the main file.  Thanks!